internal/integrations/confmap.go

```diff
@@ -0,0 +1,62 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package integrations // import "github.com/elastic/opentelemetry-collector-components/internal/integrations"
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/collector/confmap"
+)
+
+const (
+	varProviderScheme = "var"
+)
+
+func newVariablesProviderFactory(variables map[string]any) confmap.ProviderFactory {
+	return confmap.NewProviderFactory(createVariablesProvider(variables))
+}
+
+func createVariablesProvider(variables map[string]any) confmap.CreateProviderFunc {
+	return func(_ confmap.ProviderSettings) confmap.Provider {
+		return &variablesProvider{variables: variables}
+	}
+}
+
+type variablesProvider struct {
+	variables map[string]any
+}
+
+func (p *variablesProvider) Retrieve(ctx context.Context, uri string, _ confmap.WatcherFunc) (*confmap.Retrieved, error) {
+	key := strings.TrimPrefix(uri, varProviderScheme+":")
+	value, found := p.variables[key]
+	if !found {
+		return nil, fmt.Errorf("variable %q not found", key)
+	}
+
+	return confmap.NewRetrieved(value)
+}
```

Comment:
Any chance you could use unit variables here? I don't want to block this PR on this, but unit tests also help a lot with reading the code :)

Reply:
Unit tests added for this provider, let me know if any other clarification is needed.